ext/canvas/image_ops.rs

```diff
@@ -0,0 +1,698 @@
+//   let (r, g, b) = (
+//     linear_to_display_p3(r),
+//     linear_to_display_p3(g),
+//     linear_to_display_p3(b),
+//   );
+
+//   // unnormalize the value from 0.0 - 1.0
+//   (
+//     unnormalize_value_from_0_1(r),
+//     unnormalize_value_from_0_1(g),
+//     unnormalize_value_from_0_1(b),
+//   )
+// }
+
+// trait ColorSpaceConversion {
+//   /// Display P3 Color Encoding (v 1.0)
+//   /// https://www.color.org/chardata/rgb/DisplayP3.xalter
+//   fn srgb_to_display_p3(&self) -> Self;
+// }
+
+// impl<T: Primitive> ColorSpaceConversion for Rgb<T> {
+//   fn srgb_to_display_p3(&self) -> Self {
+//     let (r, g, b) = (self.0[0], self.0[1], self.0[2]);
+
+//     let (r, g, b) = apply_conversion_matrix_srgb_to_display_p3(r, g, b);
+
+//     Rgb([r, g, b])
+//   }
+// }
+
+// impl<T: Primitive> ColorSpaceConversion for Rgba<T> {
+//   fn srgb_to_display_p3(&self) -> Self {
+//     let (r, g, b, a) = (self.0[0], self.0[1], self.0[2], self.0[3]);
+
+//     let (r, g, b) = apply_conversion_matrix_srgb_to_display_p3(r, g, b);
+
+//     Rgba([r, g, b, a])
+//   }
+// }
+
+// fn process_srgb_to_display_p3<I, P, S>(image: &I) -> ImageBuffer<P, Vec<S>>
+// where
+//   I: GenericImageView<Pixel = P>,
+//   P: Pixel<Subpixel = S> + ColorSpaceConversion + 'static,
+//   S: Primitive + 'static,
+// {
+//   let (width, height) = image.dimensions();
+//   let mut out = ImageBuffer::new(width, height);
+
+//   for (x, y, pixel) in image.pixels() {
+//     let pixel = pixel.srgb_to_display_p3();
+
+//     out.put_pixel(x, y, pixel);
+//   }
+
+//   out
+// }
+
+// /// Convert the color space of the image from sRGB to Display-P3.
+// fn srgb_to_display_p3(
+//   image: DynamicImage,
+//   unmatch_color_handler: fn(
+//     ColorType,
+//     DynamicImage,
+//   ) -> Result<DynamicImage, AnyError>,
+// ) -> Result<DynamicImage, AnyError> {
+//   match image.color() {
+//     // The conversion of the lumincance color types to the display-p3 color space is meaningless.
+//     ColorType::L8 => Ok(DynamicImage::ImageLuma8(image.into_luma8())),
+//     ColorType::L16 => Ok(DynamicImage::ImageLuma16(image.into_luma16())),
+//     ColorType::La8 => Ok(DynamicImage::ImageLumaA8(image.into_luma_alpha8())),
+//     ColorType::La16 => {
+//       Ok(DynamicImage::ImageLumaA16(image.into_luma_alpha16()))
+//     }
+//     ColorType::Rgb8 => Ok(DynamicImage::ImageRgb8(process_srgb_to_display_p3(
+//       image.as_rgb8().unwrap(),
+//     ))),
+//     ColorType::Rgb16 => Ok(DynamicImage::ImageRgb16(
+//       process_srgb_to_display_p3(image.as_rgb16().unwrap()),
+//     )),
+//     ColorType::Rgba8 => Ok(DynamicImage::ImageRgba8(
+//       process_srgb_to_display_p3(image.as_rgba8().unwrap()),
+//     )),
+//     ColorType::Rgba16 => Ok(DynamicImage::ImageRgba16(
+//       process_srgb_to_display_p3(image.as_rgba16().unwrap()),
+//     )),
+//     x => unmatch_color_handler(x, image),
+//   }
+// }
+
+#[cfg(test)]
+mod tests {
```

Comment:
It would be better to add a test for EXIF orientation

Reply:
Thanks!
I added some unit tests that based on wpt.
It seems that `image` crate will soon release an API for image orientation, so we will be able to pass tests correctly with using it.

- https://github.com/image-rs/image/blob/bd0f7451a367de7ae3d898dcf1e96e9d0a1c4fa1/CHANGES.md#version-0253
- https://github.com/image-rs/image/issues/1958

Reply:
BTW, it seems that there are various difficulties with `imageOrientation`.

- https://github.com/whatwg/html/issues/8085
- https://github.com/whatwg/html/issues/7210
- https://github.com/whatwg/html/issues/8118

Reply:
I added the feature with the new released `apply_orientation` API.

- https://docs.rs/image/0.25.4/image/enum.DynamicImage.html#method.apply_orientation

The behaviour of applying `imageOrientation` with the `ImageBitmapSource` contains Exif is not clarified in the spec but in MDN currently, as I wrote at the comments in the code.

- https://github.com/mdn/content/pull/34366